### ISSUE TITLE
Scala Question

### ISSUE THREAD

**Issue by dbeavon** [databrickslabs/jupyterlab-integration#15]
Can you please clarify how the notebook experience would work if I used scala?  
I've read the following, and had follow-up questions...

https://github.com/databrickslabs/jupyterlab-integration/blob/master/docs/v2/how-it-works.md

Based on my understanding of that article, the scala kernel, a JVM, would never run locally on my workstation.  Is that correct?  It sounds everything I'm doing in each cell is being proxied to the remote cluster, including any logic that would otherwise be executed on the spark driver.

I am pretty excited by your demo, that I saw here:
https://github.com/databrickslabs/jupyterlab-integration/blob/master/docs/v2/news/scala-magic.md

I guess the concern I have is that if the scala kernel is never running on the local machine, then it will be difficult to achieve a rich scala development experience within Jupyter.  I think you highlighted some of the limitations already.  As-of now I've been using almond-sh ( https://almond.sh/ ) as my scala kernel in Jupyter and it sounds like this jupyterlab-integration experience would be very different. 

Please let me know.  I'm very eager to develop scala notebooks in Jupyterlab that will interact with a remote databricks cluster (via db-connect).  It seems like a good combination to use jupyterlab for development, along with a remote cluster that I don't need to manage myself.

**Comment by bernhard-42** [comment]
Sorry for the late reply, been on vacation

This project is tailored around Python and around the central idea of the kernel NOT running on the local machine but on the remote driver (there are many reasons for wanting this). 

The scala magic uses a trick, it send the scala code via the Context/Command REST API (https://docs.databricks.com/dev-tools/api/1.2/index.html#execution-context) to the cluster. This is just a little helper, if something can only be achieved with Scala.

**Comment by dbeavon** [comment]
Welcome back from vacation.  I wish I could have three weeks off work.  It would be nice to be a European. ;) 

We spoke on the phone right before you left.  I had (have) a case open with azure-databricks asking if it is possible to develop scala notebooks on my local workstation.  I was able to get this going to a limited extent using the "almond kernel" and JupyterLab .... but it is not the full experience at all.  For example, there is no way for the notebook to serialize its own data types and functions, so they are executed out on the remote cluster.  So I can only make use of types and functions in predefined jars.  

I see that this project is tailored around Python and the trick you mentioned for sending code to the cluster is not quite what I'm looking for.

NOTE: I did discover that the "spark-shell", within the db-connect distribution, is fully functional.  It will effectively accomplish all the things that I wanted to do in a notebook.  It will serialize brand new types and functions to the remote cluster.  But it is not as "friendly" of a U/I to work with and I would definitely prefer to use notebooks if that were possible.

I intend to experiment with zeppelin notebooks too.  But I suspect they will run into the same limitations that I experienced with Jupyter/almond.  I don't know how it works internally, but I suspect that it is technically quite challenging to properly serialize and execute functions on the remote azure-databricks cluster.  If Databricks is not sponsoring this type of thing with scala, then I probably won't get very far on my own.
